src/Analyzers/Core/Analyzers/RemoveUnusedMembers/AbstractRemoveUnusedMembersDiagnosticAnalyzer.cs

```diff
@@ -562,33 +564,82 @@ private static bool HasSyntaxErrors(INamedTypeSymbol namedTypeSymbol, Cancellati
-                        lazyModel ??= compilation.GetSemanticModel(root.SyntaxTree);
-                        var symbol = lazyModel.GetSymbolInfo(node, cancellationToken).Symbol?.OriginalDefinition;
-                        if (symbol != null && IsCandidateSymbol(symbol))
+                        foreach (var node in docComment.DescendantNodes().OfType<TIdentifierNameSyntax>())
                         {
-                            builder.Add(symbol);
+                            lazyModel ??= compilation.GetSemanticModel(syntaxTree);
+                            var symbol = lazyModel.GetSymbolInfo(node, cancellationToken).Symbol?.OriginalDefinition;
+
+                            if (IsCandidateSymbol(symbol))
+                                builder.Add(symbol);
                         }
                     }
                 }
 
-                return builder;
-            }
+                return;
 
-            private ArrayBuilder<string> GetDebuggerDisplayAttributeArguments(INamedTypeSymbol namedTypeSymbol)
-            {
-                var builder = ArrayBuilder<string>.GetInstance();
-                AddDebuggerDisplayAttributeArguments(namedTypeSymbol, builder);
-                return builder;
+                void AddAllDocumentationComments()
+                {
+                    using var _ = ArrayBuilder<TTypeDeclarationSyntax>.GetInstance(out var stack);
+
+                    foreach (var typeDeclaration in _analyzer.GetTypeDeclarations(namedTypeSymbol, cancellationToken))
+                    {
+                        stack.Clear();
+                        stack.Push(typeDeclaration);
+
+                        while (stack.Count > 0)
+                        {
+                            var currentType = stack.Pop();
+
+                            // Add the doc comments on the type itself.
+                            AddDocumentationComments(currentType, documentationComments);
+
+                            // Walk each member
+                            foreach (var member in _analyzer.GetMembers(currentType))
+                            {
+                                if (member is TTypeDeclarationSyntax childType)
+                                {
+                                    // If the member is a nested type, recurse into it.
+                                    stack.Push(childType);
+                                }
+                                else
+                                {
+                                    // Otherwise, add the doc comments on the member itself.
+                                    AddDocumentationComments(member, documentationComments);
+                                }
+                            }
+                        }
+                    }
+                }
+
+                static void AddDocumentationComments(
+                    SyntaxNode memberDeclaration, ArrayBuilder<TDocumentationCommentTriviaSyntax> documentationComments)
```

Comment:
checking if a node has a doc comment can also be optimized.  we only ever have to look at its first token, and we can avoid doing anything expensive with that token in the majority of cases.